// apico/server.js
const express = require('express');
const expressLayouts = require('express-ejs-layouts');
const path = require('path');
const fs = require('fs');
const session = require('express-session');
const { register } = require('./locals');
const cookieParser = require('cookie-parser');



/// const __dirname = new URL(import.meta.url).pathname.slice(1) ;

const app = express();

const APP_ROOT_DIR = process.cwd();

let server = null;

/// Session 
app.set('trust proxy', 1) // trust first proxy

app.use(session({
  secret: 'secret key',
  resave: false,
  saveUninitialized: true,
  cookie: { secure: true }
}))


/// Setup cookies
app.use(cookieParser());

/// To get req.body
app.use(express.json())

/// Set statics
app.use('/static', express.static('./frontend/dist') );

// Set the template engine
app.use(expressLayouts);

app.set('layout', './frontend/shared/layouts/default' );

app.set('view engine', 'ejs');

app.set('views', `./frontend` );


function start( p_port, callback )
{
    require('./index');
    
    const port  = p_port || 3000;
    server = app.listen(port, () => {

        if( typeof callback === "function" ) {
            callback();
        }
    })
}

function restart( callback )
{
    server.close(()=>{
        start( callback );
    })
}

function requireAllLocals()
{
    /// Find all js files inside tha locals folder
    // Then require them.
  
  let locals = path.join( APP_ROOT_DIR , "locals" );

  const locals_files = fs.readdirSync( locals );

  if( locals_files ) {
    locals_files.forEach( _f=>{
          let f_name = _f.split('.')[0];
          register( f_name, require(`${locals}/${ f_name }`) );
      });
  }
}

function requireAllRoutes()
{
    /// Find all js files inside tha routes folder
    // Then require them.
  
  let routes = path.join( APP_ROOT_DIR , "routes" );

  const routes_files = fs.readdirSync( routes );

  if( routes_files ) {
    routes_files.forEach( _f=>{
          let f_name = _f.split('.')[0];
          require(`${routes}/${ f_name }`)
      });
  }
}


function requireAllMiddlewares()
{
    /// Find all js files inside tha routes folder
    // Then require them.
  
  let mws = path.join( APP_ROOT_DIR , "middlewares" );

  const mws_files = fs.readdirSync( mws );

  if( mws_files ) {
    mws_files.forEach( _f=>{
          let f_name = _f.split('.')[0];
          require(`${mws}/${ f_name }`)
      });
  }
}

module.exports =  {
    start,
    restart,
    requireAllRoutes, 
    requireAllLocals,
    requireAllMiddlewares,
    app
}
